test/test-html.js

```diff
@@ -0,0 +1,273 @@
+ * @param {string} path The path of the feature.
+ * @param {Logger} logger The logger.
+ * @param {string[]} [specialElements] Dynamically allowed elements
+ */
+function lintHTMLString(html, property, path, logger, specialElements) {
+  const regexp = new RegExp(ELEMENT_REGEXP, 'gu');
+  const allowedElements = [...ALLOWED_GLOBAL_ELEMENTS];
+
+  if (specialElements) {
+    Array.prototype.push.apply(allowedElements, specialElements);
+  }
+
+  if (ALLOWED_PROPERTY_ELEMENTS[property]) {
+    Array.prototype.push.apply(
+      allowedElements,
+      ALLOWED_PROPERTY_ELEMENTS[property],
+    );
+  }
+
+  /** @type {RegExpExecArray | null} */ let match;
+  /** @type {RegExpExecArray | null} */ let attrMatch;
+
+  while (!!(match = regexp.exec(html))) {
+    const attrRegexp = new RegExp(ATTR_REGEXP, 'gu');
+    const [, actualElementName, attributesActual] = match;
+
+    const realElementName = actualElementName.toLowerCase();
+
+    if (!allowedElements.includes(realElementName)) {
+      logger.error(
+        chalk`{red {bold ${path}} - Element <${realElementName}> is not allowed in property '${property}'.}`,
+      );
+      continue;
+    }
+
+    if (actualElementName !== realElementName) {
+      logger.error(
+        chalk`{red {bold ${path}} - Use lowercase element name ({yellow <${actualElementName}>} → {green <${realElementName}>}).}`,
+      );
+    }
+
+    /** @type {string[]} */
+    const allowedAttributes = [];
+    if (ALLOWED_ELEMENT_ATTRIBUTES[realElementName]) {
+      Array.prototype.push.apply(
+        allowedAttributes,
+        ALLOWED_ELEMENT_ATTRIBUTES[realElementName],
+      );
+    }
+
+    /** @type {AttributeDescriptor[]} */
+    const badAttributes = [];
+
+    if (attributesActual) {
+      while (!!(attrMatch = attrRegexp.exec(attributesActual))) {
+        const [attrFull, attrName, attrValue] = attrMatch;
+        const attrStart = attrMatch.index;
+
+        /** @type {AttributeDescriptor} */
+        let attrDescriptor = {
+          name: attrName,
+          nameStart: attrStart,
+          nameEnd: attrStart + attrName.length,
+        };
+
+        if (attrValue != null) {
+          attrDescriptor.value = attrValue;
+          attrDescriptor.valueEnd = attrStart + attrFull.length;
+        }
+
+        if (!allowedAttributes.includes(attrName)) {
+          badAttributes.push(attrDescriptor);
+        }
+      }
+    }
+
+    if (badAttributes.length > 0) {
+      logger.error(
+        chalk`{red {bold ${path}} - Element <${realElementName}> has disallowed attributes: ${badAttributes
+          .reduce(
+            (badAttrs, { name, value }) => {
+              let result = name;
+              if (typeof value === 'string') {
+                result +=
+                  '=' + value.includes("'") ? `"${value}"` : `'${value}'`;
+              }
+              badAttrs.push(result);
+              return badAttrs;
+            },
```

Comment:
This is mostly a styling preference, but I feel that anything that isn't more than a single line should be processed outside of string interpolation and thrown into a variable, which is then referenced.  Mind if we pull this `badAttributes.reduce(...)` out into its own variable?